src/2024h2/min_generic_const_arguments.md

```diff
@@ -0,0 +1,83 @@
+# Minimum generic const arguments
+
+| Metadata | |
+| --- | --- |
+| Owner(s) | BoxyUwU |
+| Teams | lang |
+| Status | WIP |
+
+## Motivation
+
+`min_const_generics` was stabilized with the restriction that const-generic arguments may not use generic parameters other than a bare const parameter, e.g. `Foo<N>` is legal but not `Foo<{ T::ASSOC }>`. This restriction is lifted under `feature(generic_const_exprs)` however its design is fundamentally flawed and introduces significant complexity to the compiler. A ground up rewrite of the feature with a significantly limited scope (e.g. `min_generic_const_args`) would give a viable path to stabilization and result in large cleanups to the compiler.
+
+### The status quo
+
+A large amount of rust users run into the `min_const_generics` limitation that it is not legal to use generic parameters with const generics. It is generally a bad user experience to hit a wall where a feature is unfinished, and this limitation also prevents patterns that are highly desirable. We have always intended to lift this restriction since stabilizing `min_const_generics` but we did not know how.
+
+It is possible to use generic parameters with const generics by using `feature(generic_const_exprs)`. Unfortunately this feature has a number of fundamental issues that are hard to solve and as a result is *very* broken. It being so broken results in two main issues:
+- When users hit a wall with `min_const_generics` they cannot reach for the `generic_const_exprs` feature because it is either broken or has no path to stabilization.
+- In the compiler, to work around the fundamental issues with `generic_const_exprs`, we have a number of hacks which negatively affect the quality of the codebase and the general experience of contributing to the type system.
+
+### The next few steps
+
+In no particular order:
+- Author a document outlining everything wrong with `generic_const_exprs` as a sort of "design retrospective"
+- Implement a `min_generic_const_args` that would be ready for stabilization pending an accepted RFC (and potentially the new trait solver being stable)
+- Remove all hacks in the compiler relating to `generic_const_exprs`
+- Write an RFC for `min_generic_const_args`'s design once we know for certain it is a coherent design
+
+### The "shiny future" we are working towards
+
+The larger goal here is to lift most of the restrictions that const generics currently have:
```

Comment:
Part of me is wondering whether this should be a more general goal, such as "Const Generics parity with Type Generics" and throw in all of the stuff from "the larger goal". I'm a bit worried about putting too much on our plate even though it would not surprise me if some of these wind up getting stabilized throughout this goal.

Reply:
I think its best to keep scope down as much as possible so nobody feels bad if at the end of the year we've "only" got one big thing done.